apps/prairielearn/elements/pl-multiple-choice/pl-multiple-choice.py

```diff
@@ -253,91 +303,172 @@ def prepare(element_html, data):
+        "fixed-order",
+        "inline",
+        "hide-letter-keys",
+        "none-of-the-above",
+        "none-of-the-above-feedback",
+        "all-of-the-above",
+        "all-of-the-above-feedback",
+        "external-json",
+        "external-json-correct-key",
+        "external-json-incorrect-key",
+        "sort",
+        "display",
+        "hide-score-badge",  # TODO maybe rename this???
+        "allow-blank",
+    ]
+    pl.check_attribs(element, required_attribs, optional_attribs)
+
+    # Before going to the trouble of preparing answers list, check for name duplication
+    name = pl.get_string_attrib(element, "answers-name")
+
+    if name in data["params"]:
+        raise ValueError(f"Duplicate params variable name: {name}")
+    if name in data["correct_answers"]:
+        raise ValueError(f"Duplicate correct_answers variable name: {name}")
+
+    correct_answers, incorrect_answers = categorize_options(element, data)
+
+    # Check for duplicate answers. Ignore trailing/leading whitespace
+    # Making a conscious choice *NOT* to apply .lower() to all list elements in case
+    # instructors want to explicitly have matrix M vs. vector m as possible options.
+
+    choices_dict = Counter(
+        choice.html.strip() for choice in chain(correct_answers, incorrect_answers)
+    )
+
+    duplicates = [item for item, count in choices_dict.items() if count > 1]
+
+    if duplicates:
+        raise ValueError(
+            f"pl-multiple-choice element has duplicate choices: {duplicates}"
+        )
+
+    # Get answers to display to student, using a helper function to separate out logic.
+    answers_to_display = prepare_answers_to_display(
+        correct_answers,
+        incorrect_answers,
+        number_answers=pl.get_integer_attrib(element, "number-answers", None),
+        aota=get_nota_aota_attrib(
+            element, "all-of-the-above", ALL_OF_THE_ABOVE_DEFAULT
+        ),
+        nota=get_nota_aota_attrib(
+            element, "none-of-the-above", NONE_OF_THE_ABOVE_DEFAULT
+        ),
+        aota_feedback=pl.get_string_attrib(
+            element, "all-of-the-above-feedback", FEEDBACK_DEFAULT
+        ),
+        nota_feedback=pl.get_string_attrib(
+            element, "none-of-the-above-feedback", FEEDBACK_DEFAULT
+        ),
+        sort_type=get_sort_type(element),
+        display_type=get_display_type(element),
+    )
+
+    # Write to data. Because 'All of the above' is below all the correct choice(s) when it's
+    # true, the variable correct_answer will save it as correct, and overwriting previous choice(s).
+    # NOTE: The saved correct answer is just the one that gets shown to the student, it is not used for grading
     display_answers = []
     correct_answer = None
-    for i, (index, correct, html, feedback, score) in enumerate(sampled_answers):
+    for key, answer in zip(pl.iter_keys(), answers_to_display):
+        # TODO replace with a fancy key generator function off of the web.
```

Comment:
Already did this, will remove this comment as part of changes before merging.

Reply:
Done here https://github.com/PrairieLearn/PrairieLearn/pull/8903/commits/43bdad261b930462316f8fd5e00b162f2b040978